### PR TITLE
Bug 1129222 - HTTP Edge Server

### PR DESCRIPTION
**Pull request by mreid-moz**
Initial implementation of the HTTP Edge Server as a
Sandbox encoder.

### CODE REVIEW

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+-- See https://wiki.mozilla.org/CloudServices/DataPipeline/HTTPEdgeServerSpecification
+
+require "string"
+require "table"
+require 'geoip.city'
+
+function split(str, pat)
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:11]: These functions should be scoped local

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        s, e, cap = str:find(fpat, last_end)
+    end
+    if last_end <= #str then
+        cap = str:sub(last_end)
+        table.insert(t, cap)
+    end
+    return t
+end
+
+function split_path(str)
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:33]: Why the helper function when it is only called once?

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return split(str,'[\\/]+')
+end
+
+local city_db = assert(geoip.city.open(read_config("geoip_city_db")))
+local UNK_GEO = "??"
+
+function get_geo_country()
+    local country
+    local ipaddr = read_message("Fields[X-Forwarded-For]")
+    if ipaddr ~= nil then
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:43]: Just use expressions like `if ipaddr then`

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    -- Host is the name of the HTTP endpoint the client used (such as
+    -- "incoming.telemetry.mozilla.org")
+    msg.Fields.Host = read_message("Fields[Host]")
+
+    -- Path should be of the form ^/submit/namespace/id[/extra/path/components]$
+    local path = read_message("Fields[Path]")
+    local components = split_path(path)
+
+    -- Skip this message: Not enough path components.
+    if components == nil or table.getn(components) < 3 then
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:88]: split_path always returns a table component cannot be nil

Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:88]: use #components instead of getn

File: heka/sandbox/encoders/http_edge_encoder.lua (hunk 2 above)
Comment (mreid-moz) [inline heka/sandbox/encoders/http_edge_encoder.lua:33]: I was thinking about promoting these to generally-available utils.

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    if submit ~= "submit" then
+        return -1, string.format("Invalid path prefix: '%s' in %s", submit, path)
+    end
+
+    local namespace = table.remove(components, 1)
+    -- Get namespace configuration, look up params, override Logger if needed.
+    local cfg = ns_config[namespace]
+
+    -- Skip this message: Invalid namespace
+    if cfg == nil then
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:103]: `if not cfg then` (same for the other == nil comparisons)

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    -- Skip this message: Path too long
+    if pathLength > cfg.max_path_length then
+        return -1, string.format("Path too long: %d > %d", pathLength, cfg.max_path_length)
+    end
+    -- Override Logger if specified
+    if cfg["logger"] ~= nil then msg.Logger = cfg["logger"] end
+
+    -- TODO: Doesn't seem to let you override UUID. It's probably better to
+    --       use a "special" field for this anyways.
+    --msg.Uuid = table.remove(components, 1)
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:124]: Correct auto generated as per the docs

File: heka/sandbox/encoders/http_edge_encoder.lua [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+require "string"
+require "table"
+require 'geoip.city'
+
+function split(str, pat)
+    local t = {}
+    if str == nil then
+        return t
+    end
+    local fpat = "(.-)" .. pat
```
Comment (trink) [inline heka/sandbox/encoders/http_edge_encoder.lua:16]: Don't fear the PEG :)

Comment (mreid-moz) [inline heka/sandbox/encoders/http_edge_encoder.lua:16]: I don't fear it, I just don't know what it is :)

**Pull request merged**